src/node_worker.cc

```diff
@@ -344,6 +344,11 @@ void Worker::Run() {
             static_cast<EnvironmentFlags::Flags>(environment_flags_),
             thread_id_,
             std::move(inspector_parent_handle_)));
+#ifdef NODE_V8_SHARED_RO_HEAP
```

Comment:
I think this should be ifndef instead? If the user does not turn shared RO heap off, we don't need to copy the code cache, it's fine to share the one from the main instance, since the RO heap layout is the same.

Reply:
In https://github.com/nodejs/node/pull/45942#discussion_r1068237802, you mentioned that

> […] so copying code cache from one isolate and using it on an isolate with a different (non-shared) RO heap can be unsafe.

That’s what this is supposed to guard against – what am I missing here? :)

Reply:
Oh, right, so this is copying so that there is any code cache at all in the new builtin loader. What I meant was sharing by reference in workers when `NODE_V8_SHARED_RO_HEAP` is on (so no copying), and nothing when it's off. This is sharing contents of the code cache by copying in workers when `NODE_V8_SHARED_RO_HEAP` is on, and nothing when it's off.  I guess it's less nice because there is now an extra overhead from copying, and in no-snapshot mode workers and the main thread do not share updated code cache anymore, but `BuiltinCodeCacheMap` is just a map of strings -> a bunch of pointers and no-snapshot mode is rare, so I guess that's fine..

Reply:
Ah, yeah, makes sense! This is done in 2c2084c07fc8e6ecf559ff5f2b7dd880ef3a799d, and feels pretty natural anyway since there’s already concurrency management in place. But we’re not doing an `#ifndef` case here, right?

Reply:
I think if we are willing to accept that for now, `NODE_V8_SHARED_RO_HEAP` being false means that there won't be any code cache -> it would be slower to load any builtin that's not already loaded in the snapshot, then this is good to go. Before the embedder API lands it's rare to turn `NODE_V8_SHARED_RO_HEAP` off anyway, I am fine with trying to make code cache work again with `NODE_V8_SHARED_RO_HEAP` being false under certain circumstances in future PRs (that might end up being a blocker for turning pointer compression on by default? But I don't see that happening anytime soon..)

Reply:
> because a Worker thread that would inherit a code cache from the parent doesn't know whether that parent Isolate has been created by an embedder or not, and therefore also not what that parent Isolate's RO heap layout is?

In an isolate tree where the parent and the workers don't need different isolate snapshots (I think that's the common case, even if workers need custom context snapshots), it's safe to share the code cache. The issue only comes up when workers need different isolate snapshots, which I think is pretty rare? (the isolate snapshots only contain roots like strings and symbols, or other things that don't matter even if it's shared by different applications).

> and nothing really broke, right?

It caused https://github.com/nodejs/node/issues/40832 :) The corruption is still there after the fix, but now it again needs a lot of coincidence to trigger - single character strings were not that rare to appear in an application and that's why we caught it. Currently I know the bug still applies to Turbofan assertion error message strings but then...you only ever access those strings when the whole process is doomed already 😂 Though V8 can very well add other roots that might end up causing troubles in the RO heap in the future.

Reply:
Yeah, that makes sense then… is there anything to catch mismatched RO heaps in the code cache data itself? IIRC there was some kind of hash to detect mismatched V8 flags, maybe that mechanism could cover this as well?

Reply:
I guess technically that is possible but hashing the layout of RO space could be a bit expensive (need to walk the space and exclude pointers etc.), and code caching has usage outside the snapshot (e.g. I think there has been some on-disk caching strategy in Chromium for commonly loaded libraries), so that cost might matter...